azkaban-common/src/main/java/azkaban/executor/dispatcher/CandidateComparator.java

```
@@ -0,0 +1,187 @@
+
+  /** differentiate method which will kick in when the comparator list generated an equality result for
+   *  both sides. the differentiate method will try best to make sure a stable result is returned.
+   * */
+  protected boolean differentiate(T object1, T object2){
+    if (object2 == null) return true;
+    if (object1 == null) return false;
+
+    return object1.hashCode() >= object2.hashCode();
+  }
+
+  /** function to register a factorComparator to the internal Map for future reference.
+   * @param factorComparator : the comparator object to be registered.
+   * */
+  protected void registerFactorComparator(FactorComparator<T> comparator){
+      if (null == comparator ||
+          Integer.MAX_VALUE - this.getTotalWeight() < comparator.getWeight() ) {
+        logger.info(
+            "skipping registerFactorComparator as the comaractor is null or has an invalid weight value.");
+        return;
+      }
+
+      // add or replace the Comparator.
+      this.factorComparatorList.put(comparator.getFactorName(),comparator);
+      logger.info(String.format("Factor comparator added for '%s'. Weight = '%s'",
+          comparator.getFactorName(), comparator.getWeight()));
+  }
+
+  /** function update the weight of a specific registered factorCompartor.
+   * @param factorName : the name of the registered factorComparator to adjust.
+   * @param weight:      the new weight value to be adjusted to.
+   * @return -1 if the factor doesn't exist or the weight value specified is invalid,
+   *          the original value before update otherwise.
+   * */
+  public int adjustFactorWeight(String factorName, int weight){
+    // shortcut if the input is invalid.
+    if (factorName == null ||
+        factorName == "" ||
+        weight < 0 ||
+        Integer.MAX_VALUE - this.getTotalWeight() < weight){
+      logger.info(
+          "skipping adjustFactorWeight as one or more of the input parameters are invalid");
+      return -1;
```

should we throw an exception?